docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1766 @@
+
+Since most of our users do not use static type checkers we will still need to check the
+type of most user inputs so we can give them early feedback when problems arise. Thus we
+cannot remove our current error handling just because many of these errors could now be
+caught by static analysis.
+
+We can investigate using `jaxtyping`'s pytest hooks to enable runtime typecheckers like
+beartype during testing but it is not a priority for now.
+
+## Changes in documentation
+
+All type information in docstrings will be removed.
+
+Whenever there are now multiple ways of doing things, we show the ones that support
+autocomplete and static analysis most prominently. We can achieve this via tabs, similar
+to how
+[pytask](https://pytask-dev.readthedocs.io/en/stable/tutorials/defining_dependencies_products.html#products)
+does it.
+
+The general structure of the documentation is not affected by this enhancement proposal.
+
+(aligning-names)=
+
+## Aligning names
+
+### Suggested changes
+
+| **Old Name**                               | **Proposed Name**         | **Source** |
+| ------------------------------------------ | ------------------------- | ---------- |
+| `criterion`                                | `fun`                     | scipy      |
+| `criterion_kwargs`                         | `fun_kwargs`              |            |
+| `derivative`                               | `jac`                     | scipy      |
+| `derivative_kwargs`                        | `jac_kwargs`              |            |
+| `criterion_and_derivative`                 | `fun_and_jac`             |            |
+| `criterion_and_derivative_kwargs`          | `fun_and_jac_kwargs`      |            |
+| `stopping_max_criterion_evaluations`       | `stopping_maxfun`         | scipy      |
+| `stopping_max_iterations`                  | `stopping_maxiter`        | scipy      |
+| `convergence_absolute_criterion_tolerance` | `convergence_ftol_abs`    | NlOpt      |
+| `convergence_relative_criterion_tolerance` | `convergence_ftol_rel`    | NlOpt      |
+| `convergence_absolute_params_tolerance`    | `convergence_xtol_abs`    | NlOpt      |
+| `convergence_relative_params_tolerance`    | `convergence_xtol_rel`    | NlOpt      |
+| `convergence_absolute_gradient_tolerance`  | `convergence_gtol_abs`    | NlOpt      |
+| `convergence_relative_gradient_tolerance`  | `convergence_gtol_rel`    | NlOpt      |
+| `convergence_scaled_gradient_tolerance`    | `convergence_gtol_scaled` |            |
+
+### Things we do not want to align
+
+- We do not want to rename `algorithm` to `method` because our algorithm names are
+  different from scipy, so people who switch over from scipy need to adjust their code
+  anyways.
+- We do not want to rename `algo_options` to `options` for the same reason.
+- I am not sure if `params` should be renamed to `x0`. It would align estimagic more
+  with scipy, but `params` is just easier to pronounce and use as a word than `x0`.
```

Comment:
I agree

Reply:
Same here. Furthermore, to me `x` / `x0` gives the impression of a vector, not an arbitrary pytree.

Reply:
Thanks, then we have a decision. I clarified it in the text.